Port TH/THC implementation of unfold to ATen.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28597 Kill _th_zero binding, just use a simple native function instead.
* #28596 Kill THTensor_(match), which isn't used.
* #28595 Write out some set_ overloads instead of relying on code binding generation.
* #28514 Kill zero_dim_tensor_only codegen, it's not used anymore.
* #28513 Remove zero_dim_tensors only from _th_masked_fill_.
* #28512 Remove zero_dim_tensors only from _th_index_fill_.
* #28511 Kill _th_fill binding, which isn't used anymore.
* #28510 unfold: turn off device_guard
* **#28475 Port TH/THC implementation of unfold to ATen.**
* #28462 Fix scalar handling of unfold.

Differential Revision: [D18074672](https://our.internmc.facebook.com/intern/diff/D18074672)

gchanan committed:
Port TH/THC implementation of unfold to ATen.

[ghstack-poisoned]

gchanan committed:
Update on "Port TH/THC implementation of unfold to ATen."


Differential Revision: [D18074672](https://our.internmc.facebook.com/intern/diff/D18074672)

[ghstack-poisoned]

gchanan committed:
Update on "Port TH/THC implementation of unfold to ATen."


Differential Revision: [D18074672](https://our.internmc.facebook.com/intern/diff/D18074672)

[ghstack-poisoned]

albanD reviewed (approved):
LGTM

ifedan commented:
ProcessGroupGlooTest.test_allreduce_checks failed on recmdeb. Is it related to your change?

gchanan committed:
Update on "Port TH/THC implementation of unfold to ATen."


Differential Revision: [D18074672](https://our.internmc.facebook.com/intern/diff/D18074672)

[ghstack-poisoned]
